### PR TITLE
Disputes High-level rewrite & Disputes runtime 

### PR DESCRIPTION
**Pull request by rphmeier**
Closes https://github.com/paritytech/polkadot/issues/1978

* High-level overview of disputes
* Disputes module
* `InclusionInherent` -> `ParaInherent`
* A `known_disputes` runtime API

### CODE REVIEW

File: roadmap/implementers-guide/src/runtime/parainherent.md [hunk 1]
```diff
@@ -16,13 +16,19 @@ Included: Option<()>,
 
 ## Entry Points
 
-* `inclusion`: This entry-point accepts three parameters: The relay-chain parent block header, [`Bitfields`](../types/availability.md#signed-availability-bitfield) and [`BackedCandidates`](../types/backing.md#backed-candidate).
+* `enter`: This entry-point accepts three parameters: The relay-chain parent block header, [`Bitfields`](../types/availability.md#signed-availability-bitfield) and [`BackedCandidates`](../types/backing.md#backed-candidate).
     1. Hash the parent header and make sure that it corresponds to the block hash of the parent (tracked by the `frame_system` FRAME module),
+    1. Invoke `Disputes::provide_multi_dispute_data`.
+    1. If `Disputes::is_frozen`, return and set `Included` to `Some(())`.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/parainherent.md:22]: This is if we're in "governance-only" mode from having a local dispute conclude on this fork.

File: roadmap/implementers-guide/src/runtime/parainherent.md [hunk 2]
```diff
@@ -16,13 +16,19 @@ Included: Option<()>,
...
+    1. If there are any created disputes from the current session, invoke `Inclusion::collect_disputed` with the disputed candidates. Annotate each returned core with `FreedReason::Concluded`.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/parainherent.md:23]: When disputes are initiated, we remove the block from pending availability. This allows us to roll back chains to the block before blocks are _included_ as opposed to backing. It's important to do this before processing bitfields.

File: roadmap/implementers-guide/src/runtime/parainherent.md [hunk 3]
```diff
@@ -16,13 +16,19 @@ Included: Option<()>,
...
-    1. If `Scheduler::availability_timeout_predicate` is `Some`, invoke `Inclusion::collect_pending` using it, and add timed-out cores to the free cores, annotated with `FreedReason::TimedOut`.
+    1. For each freed candidate from the `Inclusion::process_bitfields` call, invoke `Disputes::note_included(current_session, candidate)`.
+    1. If `Scheduler::availability_timeout_predicate` is `Some`, invoke `Inclusion::collect_pending` using it and annotate each of those freed cores with `FreedReason::TimedOut`.
+    1. Combine and sort the dispute-freed cores, the bitfield-freed cores, and the timed-out cores.
     1. Invoke `Scheduler::clear`
-    1. Invoke `Scheduler::schedule(freed, System::current_block())`
+    1. Invoke `Scheduler::schedule(freed_cores, System::current_block())`
     1. Extract `parent_storage_root` from the parent header,
+    1. If `Disputes::could_be_invalid(current_session, candidate)` is true for any of the `backed_candidates`, fail.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/parainherent.md:31]: And we don't accept parablocks that have open disputes or disputes that have concluded against the candidate. This is notably called after initiating disputes within the current block.

File: roadmap/implementers-guide/src/runtime/parainherent.md (hunk 2 above)
Comment (rphmeier) [reply roadmap/implementers-guide/src/runtime/parainherent.md:23]: `Inclusion::collect_disputed` is kind of expensive so it's important to gate this on whether there are actually any new disputes. Which should be never.

**Comment by rphmeier** [comment]
Hm, I notice this doesn't yet handle timeouts. But I can add that.

File: roadmap/implementers-guide/src/SUMMARY.md [hunk 4]
```diff
@@ -30,6 +30,7 @@
   - [Validation Code](runtime-api/validation-code.md)
   - [Candidate Pending Availability](runtime-api/candidate-pending-availability.md)
   - [Candidate Events](runtime-api/candidate-events.md)
+  - [Known Disputes](runtime-apie/known-disputes.md)
```
Comment (drahnr) [inline roadmap/implementers-guide/src/SUMMARY.md:33]:
```suggestion
  - [Known Disputes](runtime-api/known-disputes.md)
```

File: roadmap/implementers-guide/src/runtime-apie/known-disputes.md [hunk 5]
```diff
@@ -0,0 +1 @@
+# Known Disputes
```
Comment (ordian) [inline roadmap/implementers-guide/src/runtime-apie/known-disputes.md:1]: should probably remove this file and folder

File: roadmap/implementers-guide/src/runtime/parainherent.md (hunk 2 above)
Comment (burdges) [reply roadmap/implementers-guide/src/runtime/parainherent.md:23]: both make sense

File: roadmap/implementers-guide/src/runtime/disputes.md [hunk 6]
```diff
@@ -17,22 +17,15 @@ We account for these requirements by having the disputes module handle two kinds
 1. Local disputes: those contesting the validity of the current fork by disputing a parablock included within it.
 1. Remote disputes: a dispute that has partially or fully resolved on another fork which is transplanted to the local fork for completion and eventual slashing.
 
-When handling disputes on one chain,
+When a local dispute concludes negatively, the chain needs to be abandoned and reverted back to a block where the state does not contain the bad parablock. We expect that due to the [Approval Checking Protocol](../protocol-approval.md), the current executing block should not be finalized. So we do two things when a local dispute concludes negatively:
+1. Freeze the state of parachains so nothing further is backed or included.
+1. Issue a digest in the header of the block that signals to nodes that this branch of the chain is to be abandoned.
```
Comment (burdges) [inline roadmap/implementers-guide/src/runtime/disputes.md:22]: I suppose this points back to the last ancestor that did not include/run the bad parablock.

File: roadmap/implementers-guide/src/runtime/disputes.md [hunk 7]
```diff
@@ -17,22 +17,15 @@ We account for these requirements by having the disputes module handle two kinds
...
 
-## Storage
-
-Helpers:
+If, as is expected, the chain is unfinalized, the freeze will have no effect as no honest validator will attempt to build on the frozen chain. However, if the approval checking protocol has failed and the bad parablock is finalized, the freeze serves to put the chain into a governance-only mode.
```
Comment (burdges) [inline roadmap/implementers-guide/src/runtime/disputes.md:24]: Should governance-only mode still roll back to the last unfinalized block?  I suppose we're just in such a mess we'll make humans sort it out and not bother with code for that.  :P

File: roadmap/implementers-guide/src/runtime/disputes.md [hunk 8]
```diff
@@ -66,6 +65,7 @@ PostConclusionAcceptancePeriod: BlockNumber;
   1. Fail if any disputes in the set are duplicate or concluded before the `PostConclusionAcceptancePeriod` window relative to now.
   1. Pass on each dispute statement set to `provide_dispute_data`, propagating failure.
   1. Return a list of all candidates who just had disputes initiated.
+
 * `provide_dispute_data(DisputeStatementSet) -> bool`: Provide data to an ongoing dispute or initiate a dispute.
   1. All statements must be issued under the correct session for the correct candidate.
```
Comment (burdges) [inline roadmap/implementers-guide/src/runtime/disputes.md:76]: We'll eventually need some framework for testing ousted validators acting in their now-ended session where they were still a validator.  Ugh.

**Comment by burdges** [comment]
LGTM

File: roadmap/implementers-guide/src/runtime/parainherent.md (hunk 2 above)
Comment (drahnr) [reply roadmap/implementers-guide/src/runtime/parainherent.md:23]: Could the ratiionales be added to the document?

File: roadmap/implementers-guide/src/types/disputes.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
+# Disputes
+
+## DisputeStatementSet 
+
+```rust
+/// A set of statements about a specific candidate.
+struct DisputeStatementSet {
+    candidate_hash: CandidateHash,
+    session: SessionIndex,
+    statements: Vec<(DisputeStatement, ValidatorIndex, ValidatorSignature)>,
```
Comment (drahnr) [inline roadmap/implementers-guide/src/types/disputes.md:10]: Could we not use `SignedStatement` types for this?

**Review by drahnr — APPROVED** [review]
A few nits, looks good.

File: roadmap/implementers-guide/src/runtime/disputes.md (hunk 6 above)
Comment (rphmeier) [reply roadmap/implementers-guide/src/runtime/disputes.md:22]: Yep!

File: roadmap/implementers-guide/src/runtime/disputes.md (hunk 7 above)
Comment (rphmeier) [reply roadmap/implementers-guide/src/runtime/disputes.md:24]: I'll address this in a later PR

Comment (rphmeier) [reply roadmap/implementers-guide/src/runtime/disputes.md:24]: But I don't know if it helps TBH

File: roadmap/implementers-guide/src/runtime/disputes.md (hunk 8 above)
Comment (rphmeier) [reply roadmap/implementers-guide/src/runtime/disputes.md:76]: Oh you mean validators who have been silenced? Good point...

File: roadmap/implementers-guide/src/types/disputes.md (hunk 9 above)
Comment (rphmeier) [reply roadmap/implementers-guide/src/types/disputes.md:10]: No, they're different. And `Signed<T>` is unwieldy for the reconstruction of statements.

**Comment by rphmeier** [comment]
bot merge

**Pull request merged**